### PR TITLE
Accessor improvements

### PR DESCRIPTION
**Pull request by Wuestengecko**
For details, please see the individual commits.

### CODE REVIEW

**Review by ewuerger — CHANGES_REQUESTED** [review]
It always amazes me how the simple, short and therefore future-proof solutions you come up with have so much leverage: `DeprecatedAccessor`. I know that this one will be a great weapon in the battles (releases) of the future.

Paying attention to the documentation, I also noticed that we have long hyperlinks for the type annotations that come from **autodoc** [in PR#7018](https://github.com/sphinx-doc/sphinx/pull/7018) (**not** napoleon (!) ). Since we have a very structured code base with many submodules I want to shorten these.

The only thing that keeps you from merging this is resolving my suggestions.

File: tests/test_model_layers.py [hunk 1]
```diff
@@ -449,11 +449,25 @@ def test_constraint_specification_has_linked_object_name_in_body(
     )
 
 
-def test_function_is_available_in_state(model: MelodyModel) -> None:
-    function = model.by_uuid("957c5799-1d4a-4ac0-b5de-33a65bf1519c")
-    states = function.available_in_states
-    assert "Open" in states.by_name
-    assert len(states) == 1
+class TestAttrProxyAccessor:
+    @staticmethod
```
[the remark this answers is not in the document]
Comment (Wuestengecko) [reply tests/test_model_layers.py:453]: `@staticmethod` here only tells pylint that we don't need the `self` argument, so it won't complain about it missing. pytest (which runs this) doesn't care either way.

File: tests/test_model_layers.py [hunk 2]
```diff
@@ -449,11 +449,25 @@ def test_constraint_specification_has_linked_object_name_in_body(
...
+    def test_available_in_states_can_be_modified(
+        model: capellambse.MelodyModel,
+    ):
+        function = model.by_uuid("957c5799-1d4a-4ac0-b5de-33a65bf1519c")
+        new_state = model.by_uuid("53cab5f0-fe2f-4553-8223-fbe5ea9e4d42")
+        assert len(function.available_in_states) == 1
+
+        function.available_in_states.append(new_state)
+
+        assert len(function.available_in_states) == 2
```
[the remark this answers is not in the document]
Comment (Wuestengecko) [reply tests/test_model_layers.py:470]: > I find it hilarious that I don't need to explicitly create a shallow copy

I don't think we should rely on this, see #176.

File: capellambse/model/common/accessors.py [hunk 3]
```diff
@@ -698,54 +730,28 @@ def __init__(
...
-            :class:`elements.DecoupledElementList`.  It will be used to return a
-            list of all matched objects.  Incompatible with ``xtypes =
-            None``.
+            returned. If not None, must be a subclass of
+            :class:`elements.ElementList`. It will be used to return a
+            list of all matched objects.
         """
+        del class_
         super().__init__(element.GenericElement, aslist=aslist)
         self.attr = attr
```
[the remark this answers is not in the document]
Comment (Wuestengecko) [reply capellambse/model/common/accessors.py:746]: It wasn't used here before anyways (in fact there was a warning when it was set). This change makes it at least somewhat useful, by repurposing it for static type hints.

File: tests/test_model_layers.py [hunk 4]
```diff
@@ -449,11 +449,25 @@ def test_constraint_specification_has_linked_object_name_in_body(
...
-    states = function.available_in_states
-    assert "Open" in states.by_name
-    assert len(states) == 1
+class TestAttrProxyAccessor:
+    @staticmethod
+    def test_function_is_available_in_state(model: capellambse.MelodyModel):
+        function = model.by_uuid("957c5799-1d4a-4ac0-b5de-33a65bf1519c")
+        states = function.available_in_states
+        assert "Open" in states.by_name
+        assert len(states) == 1
```
[the remark this answers is not in the document]
Comment (Wuestengecko) [reply tests/test_model_layers.py:458]: I didn't actually change these functions in any way, I only saw an opportunity to group them together. But good point, while we're here anyway :)

File: capellambse/model/common/accessors.py [hunk 5]
```diff
@@ -764,10 +770,36 @@ def __set__(
...
+        if value._model is not elmlist._parent._model:
+            raise ValueError("Cannot insert elements from different models")
+        objs = [*elmlist[:index], value, *elmlist[index:]]
+        self.__set_links(elmlist._parent, objs)
+
+    def delete(
+        self, elmlist: ElementListCouplingMixin, obj: element.ModelObject
+    ) -> None:
+        assert self.aslist is not None
+        objs = [i for i in elmlist if i != obj]
```
[the remark this answers is not in the document]
Comment (Wuestengecko) [reply capellambse/model/common/accessors.py:791]: No, those would trigger a `delete` again and result in an infinite loop (or an exception, not sure).

**Review by ewuerger — APPROVED** [review]
LGTM

**Pull request merged**